### ISSUE TITLE
Pix2Pix tutorial BatchNorm issue

### ISSUE THREAD

**Issue by EmanueleGhelfi** [tensorflow/tensorflow#29478]
## URL(s) with the issue:
https://www.tensorflow.org/alpha/tutorials/generative/pix2pix

## Description of issue (what needs changing):

The `generate_images` function:

```python
def generate_images(model, test_input, tar):
  # the training=True is intentional here since
  # we want the batch statistics while running the model
  # on the test dataset. If we use training=False, we will get
  # the accumulated statistics learned from the training dataset
  # (which we don't want)
  prediction = model(test_input, training=True)
  plt.figure(figsize=(15,15))

  display_list = [test_input[0], tar[0], prediction[0]]
  title = ['Input Image', 'Ground Truth', 'Predicted Image']

  for i in range(3):
    plt.subplot(1, 3, i+1)
    plt.title(title[i])
    # getting the pixel values between [0, 1] to plot it.
    plt.imshow(display_list[i] * 0.5 + 0.5)
    plt.axis('off')
  plt.show()
```

generate images using the generator model with the flag training=True. The problem is that in this way the batch normalization statistics (moving mean and moving variance) are updated using the test set statistics. 

This is wrong. The original pix2pix paper asserts that they evaluate the model using the flag training=True but they do this in order to normalize using the minibatch (with batch size = 1) statistics. This is done only in the test phase and statistics should not be saved into the model.

I think that a better approach is to visualize the data generated during training (not re-calling the generator but using the data generated in order to calculate the loss). 

Once the training is finished we can evaluate the model.

The problem is that in tf 2.0 is not possible to use the minibatch statistics in the batch normalization layer. Every time we call BatchNorm()(input, training=True) the moving mean and moving variance are updated. I think that this can be managed in a better way by adding a flag that tells the layer whether to use the minibatch statistics.

**Comment by EmanueleGhelfi** [comment]
Actually this is not an issue since the batch norm by default uses the statistics of the current batch during training. Since we never use the model with `training=False` this is not a problem.

**Comment by fengwang** [comment]
@EmanueleGhelfi This does matter. In every epoch, after making such a prediction, all the batch normalization layer are updated with the information from `test_input`, which is not the desired behavior.

**Comment by EmanueleGhelfi** [comment]
But since pix2pix always use `training=True` it always use batch statistics. So the moving mean and variance are not used anywhere.

**Comment by fengwang** [comment]
@EmanueleGhelfi With the code below

```python
from tensorflow.python.keras.layers import Input
from tensorflow.python.keras.models import Model
from tensorflow.python.keras.layers import BatchNormalization
from tensorflow.python.keras.layers import Conv2D

input = Input( (3, 3, 1) )
x = Conv2D( 1, 1, )(input)
x = BatchNormalization(momentum=0.8)(x)
x = Conv2D( 1, 1, activation='sigmoid' )(x)
model = Model( input, x )
model.compile( loss='mae', optimizer='adam' )

import numpy as np
a1 = np.random.random( (1, 3, 3, 1) )
a2 = np.random.random( (1, 3, 3, 1) )
model.train_on_batch( a1, a2 )

model.save( './model_pre.h5' )

import os
os.system( '/bin/md5sum ./model_pre.h5' )

a = np.random.random( (1, 3, 3, 1) )
b = model( np.asarray(a, dtype='float32'), training=True )
model.save( './model_after.h5' )

os.system( '/bin/md5sum ./model_after.h5' )

```

I got such output by running it twice:

```
 feng@DDM5865  ~/tmp/bn  python3 ./test.py
WARNING:tensorflow:From /usr/lib/python3.7/site-packages/tensorflow/python/ops/resource_variable_ops.py:435: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
WARNING:tensorflow:From /usr/lib/python3.7/site-packages/tensorflow/python/ops/math_ops.py:3066: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
2019-06-06 14:17:17.444135: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2809000000 Hz
2019-06-06 14:17:17.444568: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x562041bffe50 executing computations on platform Host. Devices:
2019-06-06 14:17:17.444598: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
4beaa9f45f17c7db87679092080ebf8c  ./model_pre.h5
4beaa9f45f17c7db87679092080ebf8c  ./model_after.h5
 feng@DDM5865  ~/tmp/bn  python3 ./test.py
WARNING:tensorflow:From /usr/lib/python3.7/site-packages/tensorflow/python/ops/resource_variable_ops.py:435: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
WARNING:tensorflow:From /usr/lib/python3.7/site-packages/tensorflow/python/ops/math_ops.py:3066: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
2019-06-06 14:17:22.777565: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2809000000 Hz
2019-06-06 14:17:22.777901: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x56502d9f7700 executing computations on platform Host. Devices:
2019-06-06 14:17:22.777942: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
df0651ec886876f724a36ea1f5cf208b  ./model_pre.h5
cfea66f48d5e93c2203512f730cd35be  ./model_after.h5
```

And I found for the first time, the model is not updated. But for the second one, the model did learn something new. I am confused....

**Comment by EmanueleGhelfi** [comment]
@fengwang yes, the model updates the moving mean and moving variance. You can also inspect them by accessing the correct attributes inside the BatchNormalization layer. The fact is that if the flag training is True the layer uses only the current mean and variance in order to calculate its output. In this way the moving mean and variance are not updated if I'm not wrong. You can also try this by inspecting the output of the layer with respect to its mean and its variance. I'll try it.

**Comment by EmanueleGhelfi** [comment]
I confirm my comment: 
if the flag training is True the layer uses only the current mean and variance in order to calculate its output. In this way the moving mean and variance are not updated.

**Comment by yashk2810** [comment]
If this is not a problem anymore, @EmanueleGhelfi can you please close this.

If not, can you send us a PR (https://github.com/tensorflow/docs/blob/master/site/en/r2/tutorials/generative/pix2pix.ipynb). Any improvement to the docs is always appreciated :)

**Comment by EmanueleGhelfi** [comment]
Hi @yashk2810 I think the code has still problems right now. In the next days I'll try to submit a PR.

**Comment by prathamesh0** [comment]
Hi! The model gives all NaNs when used with training = False on test images. Can anyone please explain this?

**Comment by Saduf2019** [comment]
@EmanueleGhelfi 
Please post this issue on[ keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
